### ISSUE TITLE
Phase 1: Build next-token examples with separate training and validation data

### ISSUE THREAD

**Issue by Dzejrou** [Dzejrou/llm#4]
## Context

Phase 1 of [Roadmap 1](https://github.com/Dzejrou/llm/issues/1) now has a Rust library and a byte tokenizer: #2 was completed by PR #3, merged as `8b89c595653d243435b332625cddc8ed0ea69169`. `ByteTokenizer::encode` produces a `Vec<TokenId>`; `TokenId` is `u32`.

The next step is to turn those IDs into supervised next-token examples, with training and validation data kept separate. This issue completes the planned data foundation before the bigram learning loop.

## Concept

The text supplies its own labels: the correct answer at a position is the token that actually followed it. For token IDs `[10, 20, 30, 40]` and context length `T = 3`, the example is:

```text
inputs:  [10, 20, 30]
targets: [20, 30, 40]
```

Each pair of aligned positions asks for the next token. A length-`T` example needs `T + 1` source tokens. For a sequence of length `N`, taking every valid starting position produces `max(N - T, 0)` windows. Neighboring windows overlap; within a window, the input and target slices overlap too. The data layer only constructs examples; enforcing causal access inside a future Transformer is the model's responsibility.

Validation data is held out from parameter updates so its loss can later help us assess behavior on unseen data. Split the token stream into disjoint contiguous portions **before** creating windows. Splitting already-overlapping windows can put the same source positions into both training and validation. A contiguous split prevents that overlap; it does not guarantee that the original text contains no repeated passages.

## Goal

A small, deterministic data API splits a token stream and lazily produces correctly shifted next-token examples from either portion, without copying or materializing the full set of windows.

## Requirements

1. Add a focused public `data` module to the existing library. Reuse `TokenId`; keep it independent of `ByteTokenizer`, model code, and any specific vocabulary size.
2. Provide a checked contiguous split operation accepting `tokens: &[TokenId]` and an explicit `train_len: usize`:
   - Training is exactly `tokens[..train_len]`; validation is exactly `tokens[train_len..]`.
   - Reject `train_len > tokens.len()` with a descriptive error, not a panic or silent clamp.
   - Permit the endpoints `0` and `tokens.len()`, including empty input. They are valid data partitions; a future training command may impose stricter minimum-data requirements.
   - Return borrowed immutable slices, either directly or in a small named result type. Do not copy the corpus.
3. Provide next-token window iteration over a supplied slice and a positive context length `T`:
   - Each item exposes borrowed `inputs: &[TokenId]` and `targets: &[TokenId]`, both of length `T`.
   - At starting position `s`, inputs are `tokens[s..s+T]` and targets are `tokens[s+1..s+T+1]`.
   - Visit all valid starts in ascending order with stride one; emit exactly `max(N - T, 0)` examples.
   - Reject `T == 0` explicitly. Positive `T` with insufficient tokens yields an empty iterator, including empty input, `N == T`, and an oversized context such as `usize::MAX`. Avoid overflowing `T + 1` or unsigned subtraction.
   - No padding, wraparound, implicit concatenation, shuffling, or random sampling.
   - Consume only the supplied slice. In the documented workflow, construct iterators separately from the train and validation slices so no example crosses their boundary.
4. Treat token IDs as opaque integers. Do not impose the byte tokenizer's `0..255` restriction here or attempt to decode a window as UTF-8. Splits and windows may cut a multibyte character; they remain valid model data even when a slice alone is not valid text.
5. Provide useful errors with `Debug`, `Display`, and `std::error::Error`. Keep the API small; a concrete iterator and a small borrowed example type are sufficient. Standard-library facilities are enough for this issue.
6. Extend the README and public API documentation with one runnable, tested example: encode a short synthetic string, split the IDs at an explicit token index, then inspect a train window and a validation window. Document the order of operations and the empty-split/short-input behavior. Keep the existing tokenizer examples valid.

## Suggested design

Functions such as `split_tokens(tokens, train_len)` and `next_token_windows(tokens, context_len)`, plus a small example type containing the two slices, are enough. Names and iterator mechanics are left to the worker. No dataset trait, generic loader, tensor type, or owned collection of every example is needed.

Use an explicit token index for the split instead of a floating-point ratio. That makes boundaries and edge cases deterministic; any ratio-to-index policy can be introduced at the application layer when useful. Both APIs should borrow the same source buffer. Do not add a combined tokenizer/data abstraction.

## Tests

- Verify the exact example above for `[10, 20, 30, 40]` with `T = 3`.
- With a longer sequence and `T = 2`, verify every window's contents, order, and count, including the last valid window.
- With `T = 1`, produce exactly the adjacent token pairs that the upcoming bigram model will use.
- Verify `N == T + 1` gives one example; `N == T`, `N < T`, and empty input give none for positive `T`.
- Verify `T == 0` returns an error and `T == usize::MAX` on a short slice produces no windows without overflow or panic.
- Verify valid split points, both endpoints, empty input at index zero, and rejection of out-of-bounds indices, including `usize::MAX`.
- Boundary isolation: split `[0, 1, 2, 3, 4, 5, 6, 7]` at index 4 and use `T = 2`. The two training examples must use only positions 0 through 3, and the two validation examples only positions 4 through 7; the transition `3 -> 4` must not appear in either split's windows.
- Verify repeated iteration from the same input produces the same windows.
- Verify IDs above 255, including `u32::MAX`, are preserved as ordinary data.
- Verify a tokenizer-produced multibyte sequence can be partitioned/windowed inside a character without decoding or rejection. This is a data-boundary test, not a request to change strict tokenizer decoding.
- Keep existing tokenizer and documentation tests passing; exercise the new public example as a doctest.

Run `cargo fmt --check`, `cargo clippy --all-targets -- -D warnings`, and `cargo test`.

## Educational/comment requirements

Explain where targets come from, why they are shifted by one position, why `T` input positions need `T + 1` source tokens, and why the last valid starting index is `N - T - 1` when `N > T`. Explain how borrowed overlapping slices represent the examples without copying their contents. Document why train/validation splitting precedes window construction and why UTF-8 decoding is irrelevant inside this layer.

## Acceptance criteria

- [ ] The public split and window APIs implement the specified boundaries, errors, and deterministic ordering.
- [ ] Windows borrow the source tokens and do not materialize all examples in memory.
- [ ] Tests verify shifts, counts, edge cases, overflow avoidance, and train/validation isolation.
- [ ] Documentation teaches the workflow and includes a checked example.
- [ ] Formatting, Clippy, and all tests pass.
- [ ] The PR links this issue with `Closes #<this issue number>`, explains consequential choices, and records validation results.

## Out of scope

External datasets, file I/O/downloads, multiple-document packing, CLI commands, random sampling, minibatch assembly, padding/special tokens, tensors, models, loss functions, gradients, optimizers, training loops, and generation. Use small synthetic sequences in tests and examples. Preserve the existing tokenizer contract.
